tofu/geom/_GG.pyx

```diff
@@ -925,218 +883,366 @@ def discretize_vpoly(double[:,::1] VPoly, double dL,
         else:
-            #indI.append(list(range(nPhi0,NRPhi_int)+list(range(0,nPhi1+1))))
-            Phin[ii] = nPhi1+1+NRPhi_int-nPhi0
-            if ii==0:
-                indI = np.nan*np.ones((Rn,Phin[ii]*Rratio+1))
-            for jj in range(0,NRPhi_int-nPhi0):
-                indI[ii,jj] = <double>( nPhi0+jj )
-            for jj in range(NRPhi_int-nPhi0,Phin[ii]):
-                indI[ii,jj] = <double>( jj- (NRPhi_int-nPhi0) )
-        NP += Zn*Phin[ii]
-    Pts = np.empty((3,NP))
-    ind = np.empty((NP,))
-    dV = np.empty((NP,))
-    # Compute Pts, dV and ind
-    # This triple loop is the longest part, it takes ~90% of the CPU time
-    NP = 0
-    if Out.lower()=='(x,y,z)':
-        for ii in range(0,Rn):
-            # To make sure the indices are in increasing order
-            iii = np.sort(indI[ii,~np.isnan(indI[ii,:])])
-            for zz in range(0,Zn):
-                for jj in range(0,Phin[ii]):
-                    indiijj = iii[jj]
-                    phi = -Cpi + (0.5+indiijj)*dPhir[ii]
-                    Pts[0,NP] = R[ii]*Ccos(phi)
-                    Pts[1,NP] = R[ii]*Csin(phi)
-                    Pts[2,NP] = Z[zz]
-                    ind[NP] = NRPhi0[ii] + indZ[zz]*NRPhi[ii] + indiijj
-                    dV[NP] = dRr*dZr*dRPhir[ii]
-                    NP += 1
+            nphi1 = int(Cfloor((max_phi+Cpi) * inv_drphi))
+        sz_phi[0] = nphi1 + 1 - nphi0
+        max_sz_phi[0] = sz_phi[0]
+        indI = -np.ones((sz_r, sz_phi[0] * r_ratio + 1), dtype=int)
+        indi_mv = indI
+        for jj in range(sz_phi[0]):
+            indi_mv[0, jj] = nphi0 + jj
+        NP += sz_z * sz_phi[0]
     else:
-        for ii in range(0,Rn):
-            iii = np.sort(indI[ii,~np.isnan(indI[ii,:])])
-            #assert iii.size==Phin[ii] and np.all(np.unique(iii)==iii)
-            for zz in range(0,Zn):
-                for jj in range(0,Phin[ii]):
-                    indiijj = iii[jj] #indI[ii,iii[jj]]
-                    Pts[0,NP] = R[ii]
-                    Pts[1,NP] = Z[zz]
-                    Pts[2,NP] = -Cpi + (0.5+indiijj)*dPhir[ii]
-                    ind[NP] = NRPhi0[ii] + indZ[zz]*NRPhi[ii] + indiijj
-                    dV[NP] = dRr*dZr*dRPhir[ii]
-                    NP += 1
+        # Get the actual RPhi resolution and Phi mesh elements (! depends on R!)
+        ncells_rphi[0] = <int>Cceil(twopi_over_dphi * disc_r[0])
+        loc_nc_rphi = ncells_rphi[0]
+        step_rphi[0] = _TWOPI / ncells_rphi[0]
+        inv_drphi = 1. / step_rphi[0]
+        reso_phi_mv[0] = step_rphi[0] * disc_r[0]
+        tot_nc_plane[0] = 0 # initialization
+        # Get index and cumulated indices from background
+        for jj in range(ind_loc_r0, ncells_r0[0]):
+            if disc_r0[jj]==disc_r[0]:
+                ind_loc_r0 = jj
+                break
+            else:
+                ncells_rphi0 += <long>Cceil(twopi_over_dphi * disc_r0[jj])
```

Comment:
I may be misunderstanding, but from here I feel that:
* You keep the angle resolution (phistep) constant and it is the distance resolution (r*phistep) that changes, meaning that the mesh is a classic cylindrical mesh
* But we want to have the distance resolution constant (r*phistep), so the number of cells in the phi direction varies for every value of R

Could you clarify in case I am misunderstanding ?